### PR TITLE
AEA-1281 Enable stronger typing checks

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/test_tools/click_testing.py [hunk 1]
```diff
@@ -88,18 +89,18 @@ def invoke(
                 exc_info = sys.exc_info()
             finally:
                 sys.stdout.flush()
-                stdout = outstreams[0].getvalue() if not outstreams[0].closed else b""
+                stdout = outstreams[0].getvalue() if not outstreams[0].closed else b""  # type: ignore
                 if self.mix_stderr:
-                    stderr = None
+                    stderr: Optional[bytes] = None
```
Comment (DavidMinarsch) [inline aea/test_tools/click_testing.py:94]: This is strange @marcofavorito . We make it raise a ValueError. See test. Why?

File: packages/fetchai/connections/p2p_libp2p_client/connection.py [hunk 2]
```diff
@@ -316,7 +322,10 @@ async def _receive(self) -> Optional[bytes]:
             if not data:  # pragma: no cover
                 return None
             return data
-        except asyncio.streams.IncompleteReadError as e:  # pragma: no cover
+        except ConnectionError as e:
+            self.logger.info(f"Connection error: {e}")
+            return None
+        except IncompleteReadError as e:  # pragma: no cover
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p_client/connection.py:328]: reverse order

File: aea/cli/publish.py [hunk 3]
```diff
@@ -128,7 +128,9 @@ def _check_is_item_in_remote_registry(
     get_package_meta(item_type_plural[:-1], public_id)
 
 
-def _check_is_item_in_local_registry(public_id, item_type_plural, registry_path):
+def _check_is_item_in_local_registry(
+    public_id, item_type_plural, registry_path
```
Comment (solarw) [inline aea/cli/publish.py:132]: no types for the arguments?

File: aea/cli/registry/utils.py [hunk 4]
```diff
@@ -168,15 +169,15 @@ def extract(source: str, target: str) -> None:
...
+def _rm_tarfiles() -> None:
     cwd = os.getcwd()
     for filename in os.listdir(cwd):
         if filename.endswith(".tar.gz"):
             filepath = os.path.join(cwd, filename)
             os.remove(filepath)
 
 
-def clean_tarfiles(func):
+def clean_tarfiles(func) -> Callable:
```
Comment (solarw) [inline aea/cli/registry/utils.py:180]:
```suggestion
def clean_tarfiles(func: Callable) -> Callable:
```

File: aea/cli/scaffold.py [hunk 5]
```diff
@@ -225,7 +225,7 @@ def _scaffold_dm_handler(ctx: Context):
     )
 
 
-def _scaffold_error_handler(ctx):
+def _scaffold_error_handler(ctx) -> None:
```
Comment (solarw) [inline aea/cli/scaffold.py:228]:
```suggestion
def _scaffold_error_handler(ctx: Context) -> None:
```

File: aea/cli/search.py [hunk 6]
```diff
@@ -143,7 +144,7 @@ def setup_search_ctx(ctx: Context, local: bool) -> None:
         logger.debug("Using registry {}".format(registry_directory))
 
 
-def _is_invalid_item(name, dir_path, config_path):
+def _is_invalid_item(name, dir_path, config_path) -> bool:
```
Comment (solarw) [inline aea/cli/search.py:147]: no types for the arguments

File: aea/cli/search.py [hunk 7]
```diff
@@ -169,7 +170,7 @@ def _get_details_from_dir(
         results.append(details)
 
 
-def _search_items_locally(ctx, item_type_plural):
+def _search_items_locally(ctx, item_type_plural) -> List[Dict]:
```
Comment (solarw) [inline aea/cli/search.py:173]:
```suggestion
def _search_items_locally(ctx: Context, item_type_plural: str) -> List[Dict]:
```

File: aea/cli/utils/click_utils.py [hunk 8]
```diff
@@ -68,19 +68,21 @@ def arg_strip(s):
...
         """
         Initialize the Public Id parameter.
 
         Just forwards arguments to parent constructor.
         """
-        super().__init__(*args, **kwargs)
+        super().__init__(*args, **kwargs)  # type: ignore
 
-    def get_metavar(self, param):
+    def get_metavar(self, param) -> str:
```
Comment (solarw) [inline aea/cli/utils/click_utils.py:81]: no argument types

File: aea/cli/utils/click_utils.py [hunk 9]
```diff
@@ -68,19 +68,21 @@ def arg_strip(s):
...
         """Return the metavar default for this param if it provides one."""
         return "PUBLIC_ID"
 
-    def convert(self, value, param, ctx):
+    def convert(self, value, param, ctx) -> PublicId:
```
Comment (solarw) [inline aea/cli/utils/click_utils.py:85]: no argument types

File: aea/cli/utils/click_utils.py [hunk 10]
```diff
@@ -91,17 +93,17 @@ def convert(self, value, param, ctx):
...
 
-    def __init__(self):
+    def __init__(self) -> None:
         """Initialize the agent directory parameter."""
         super().__init__(
             exists=True, file_okay=False, dir_okay=True, readable=True, writable=False
         )
 
-    def get_metavar(self, param):
+    def get_metavar(self, param) -> str:
```
Comment (solarw) [inline aea/cli/utils/click_utils.py:102]: no argument types

File: aea/cli/utils/click_utils.py [hunk 11]
```diff
@@ -91,17 +93,17 @@ def convert(self, value, param, ctx):
...
         """Return the metavar default for this param if it provides one."""
         return "AGENT_DIRECTORY"  # pragma: no cover
 
-    def convert(self, value, param, ctx):
+    def convert(self, value, param, ctx) -> str:  # type: ignore
```
Comment (solarw) [inline aea/cli/utils/click_utils.py:106]: no argument types

File: aea/cli/utils/formatting.py [hunk 12]
```diff
@@ -26,7 +26,7 @@
 from aea.exceptions import enforce
 
 
-def format_items(items):
+def format_items(items) -> str:
```
Comment (solarw) [inline aea/cli/utils/formatting.py:29]: no argument types

File: aea/cli/utils/loggers.py [hunk 13]
```diff
@@ -43,28 +44,28 @@ class ColorFormatter(logging.Formatter):
         "warning": dict(fg="yellow"),
     }
 
-    def format(self, record):
+    def format(self, record) -> str:
```
Comment (solarw) [inline aea/cli/utils/loggers.py:47]: no argument types

File: aea/cli/utils/loggers.py [hunk 14]
```diff
@@ -43,28 +44,28 @@ class ColorFormatter(logging.Formatter):
...
             msg = record.getMessage()
             if level in self.colors:
-                prefix = click.style("{}: ".format(level), **self.colors[level])
+                prefix = click.style("{}: ".format(level), **self.colors[level])  # type: ignore
                 msg = "\n".join(prefix + x for x in msg.splitlines())
             return msg
         return logging.Formatter.format(self, record)  # pragma: no cover
 
 
 def simple_verbosity_option(
```
Comment (solarw) [inline aea/cli/utils/loggers.py:59]: no argument types

File: aea/configurations/data_types.py [hunk 15]
```diff
@@ -111,7 +111,7 @@ def __eq__(self, other) -> bool:
         """Check equality."""
         return isinstance(other, PackageVersion) and self._version == other._version
 
-    def __lt__(self, other):
+    def __lt__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:114]: no argument types

File: aea/configurations/data_types.py [hunk 16]
```diff
@@ -365,25 +365,25 @@ def json(self) -> Dict:
...
             author=self.author, name=self.name, version=self.version
         )
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         """Get the representation."""
         return f"<{self}>"
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:386]: no argument types

File: aea/configurations/data_types.py [hunk 17]
```diff
@@ -392,7 +392,7 @@ def __eq__(self, other):
             and self.version == other.version
         )
 
-    def __lt__(self, other):
+    def __lt__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:395]: no argument types

File: aea/configurations/data_types.py [hunk 18]
```diff
@@ -515,29 +517,29 @@ def to_uri_path(self) -> str:
...
             package_type=self.package_type.value, public_id=self.public_id,
         )
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         """Get the object representation in string."""
         return f"PackageId{self.__str__()}"
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:534]: no argument types

File: aea/configurations/data_types.py [hunk 19]
```diff
@@ -515,29 +517,29 @@ def to_uri_path(self) -> str:
...
         """Compare with another object."""
         return (
             isinstance(other, PackageId)
             and self.package_type == other.package_type
             and self.public_id == other.public_id
         )
 
-    def __lt__(self, other):
+    def __lt__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:542]: no argument types

File: aea/configurations/data_types.py [hunk 20]
```diff
@@ -757,7 +761,7 @@ def __str__(self) -> str:
         """Get the string representation."""
         return f"{self.__class__.__name__}(name='{self.name}', version='{self.version}', index='{self.index}', git='{self.git}', ref='{self.ref}')"
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/configurations/data_types.py:764]: no argument types

File: aea/configurations/validation.py [hunk 21]
```diff
@@ -92,14 +92,14 @@ def __repr__(self) -> str:  # pragma: nocover
 class CustomTypeChecker(TypeChecker):
     """Custom type checker to handle env variables."""
 
-    def is_type(self, instance, type):  # pylint: disable=redefined-builtin
+    def is_type(self, instance, type) -> bool:  # pylint: disable=redefined-builtin
```
Comment (solarw) [inline aea/configurations/validation.py:95]: no argument types

File: aea/configurations/validation.py [hunk 22]
```diff
@@ -92,14 +92,14 @@ def __repr__(self) -> str:  # pragma: nocover
...
         """Check is instance of type."""
         if is_env_variable(instance):
             return True
         return super().is_type(instance, type)
 
 
-def ownAdditionalProperties(validator, aP, instance, schema):
+def ownAdditionalProperties(validator, aP, instance, schema) -> Iterator:
```
Comment (solarw) [inline aea/configurations/validation.py:102]: no argument types

File: aea/configurations/validation.py [hunk 23]
```diff
@@ -267,15 +267,18 @@ def validate_data_with_pattern(
...
+        excludes_ = excludes
     pattern_path_value = {
         tuple(path): value for path, value in dict_to_path_value(pattern)
     }
     data_path_value = {tuple(path): value for path, value in dict_to_path_value(data)}
     errors = []
 
-    def check_excludes(path):
-        for exclude in excludes:
+    def check_excludes(path) -> bool:
```
Comment (solarw) [inline aea/configurations/validation.py:280]: no argument types

File: aea/helpers/base.py [hunk 24]
```diff
@@ -530,14 +533,14 @@ def reachable_nodes(
 class cached_property:  # pragma: nocover
     """Cached property from python3.8 functools."""
 
-    def __init__(self, func):
+    def __init__(self, func) -> None:
```
Comment (solarw) [inline aea/helpers/base.py:536]: no argument types

File: aea/helpers/base.py [hunk 25]
```diff
@@ -530,14 +533,14 @@ def reachable_nodes(
...
         """Init cached property."""
         self.func = func
         self.attrname = None
         self.__doc__ = func.__doc__
         self.lock = RLock()
 
-    def __set_name__(self, _, name):
+    def __set_name__(self, _, name) -> None:
```
Comment (solarw) [inline aea/helpers/base.py:543]: no argument types

File: aea/helpers/base.py [hunk 26]
```diff
@@ -547,7 +550,7 @@ def __set_name__(self, _, name):
                 f"({self.attrname!r} and {name!r})."
             )
 
-    def __get__(self, instance, _=None):
+    def __get__(self, instance, _=None) -> Any:
```
Comment (solarw) [inline aea/helpers/base.py:553]: no argument types

File: aea/helpers/base.py [hunk 27]
```diff
@@ -800,7 +803,7 @@ def from_json(cls, obj: Dict) -> "CertRequest":
         """Compute the JSON representation."""
         return cls(**obj)
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/base.py:806]: no argument types

File: aea/helpers/base.py [hunk 28]
```diff
@@ -834,7 +837,7 @@ def compute_specifier_from_version(version: Version) -> str:
     return specifier_set
 
 
-def decorator_with_optional_params(decorator):
+def decorator_with_optional_params(decorator) -> Callable:
```
Comment (solarw) [inline aea/helpers/base.py:840]: no argument types

File: aea/helpers/file_lock.py [hunk 29]
```diff
@@ -34,14 +34,14 @@
     LOCK_NB = win32con.LOCKFILE_FAIL_IMMEDIATELY
     __overlapped = pywintypes.OVERLAPPED()
 
-    def lock(file, flags):
+    def lock(file, flags) -> None:
```
Comment (solarw) [inline aea/helpers/file_lock.py:37]: no argument types

File: aea/helpers/file_lock.py [hunk 30]
```diff
@@ -34,14 +34,14 @@
...
         """Lock a file with flags."""
         hfile = win32file._get_osfhandle(  # pylint: disable=protected-access
             file.fileno()
         )
         win32file.LockFileEx(hfile, flags, 0, 0xFFFF0000, __overlapped)
 
-    def unlock(file):
+    def unlock(file) -> None:
```
Comment (solarw) [inline aea/helpers/file_lock.py:44]: no argument types

File: aea/helpers/file_lock.py [hunk 31]
```diff
@@ -53,11 +53,11 @@ def unlock(file):
     import fcntl
     from fcntl import LOCK_EX, LOCK_NB, LOCK_SH  # noqa # pylint: disable=unused-import
 
-    def lock(file, flags):
+    def lock(file, flags) -> None:
         """Lock a file with flags."""
         fcntl.flock(file.fileno(), flags)
 
-    def unlock(file):
+    def unlock(file) -> None:
```
Comment (solarw) [inline aea/helpers/file_lock.py:60]: no argument types

File: aea/helpers/file_lock.py [hunk 32]
```diff
@@ -53,11 +53,11 @@ def unlock(file):
     import fcntl
     from fcntl import LOCK_EX, LOCK_NB, LOCK_SH  # noqa # pylint: disable=unused-import
 
-    def lock(file, flags):
+    def lock(file, flags) -> None:
```
Comment (solarw) [inline aea/helpers/file_lock.py:56]: no argument types

File: aea/helpers/multiaddr/base.py [hunk 33]
```diff
@@ -75,17 +75,17 @@ def digest(self) -> bytes:
     )
 
 
-def _pad_scalar(scalar):
+def _pad_scalar(scalar) -> bytes:
```
Comment (solarw) [inline aea/helpers/multiaddr/base.py:78]: no argument types

File: aea/helpers/multiaddr/base.py [hunk 34]
```diff
@@ -75,17 +75,17 @@ def digest(self) -> bytes:
...
 
 
-def _pad_hex(hexed):
+def _pad_hex(hexed) -> str:
     """Pad odd-length hex strings."""
     return hexed if not len(hexed) & 1 else "0" + hexed
 
 
-def _hex_to_bytes(hexed):
+def _hex_to_bytes(hexed) -> bytes:
```
Comment (solarw) [inline aea/helpers/multiaddr/base.py:88]: no argument types

File: aea/helpers/storage/generic_storage.py [hunk 35]
```diff
@@ -103,7 +105,7 @@ async def list(self) -> List[OBJECT_ID_AND_BODY]:
 class SyncCollection:
     """Async collection."""
 
-    def __init__(self, async_collection_coro, loop: asyncio.AbstractEventLoop):
+    def __init__(self, async_collection_coro, loop: asyncio.AbstractEventLoop) -> None:
```
Comment (solarw) [inline aea/helpers/storage/generic_storage.py:108]: no argument types

File: aea/helpers/storage/generic_storage.py [hunk 36]
```diff
@@ -113,7 +115,7 @@ def __init__(self, async_collection_coro, loop: asyncio.AbstractEventLoop):
         self._loop = loop
         self._async_collection = self._run_sync(async_collection_coro)
 
-    def _run_sync(self, coro):
+    def _run_sync(self, coro) -> Any:
```
Comment (solarw) [inline aea/helpers/storage/generic_storage.py:118]: no argument types

File: aea/helpers/transaction/base.py [hunk 37]
```diff
@@ -96,15 +94,15 @@ def decode(cls, raw_transaction_protobuf_object) -> "RawTransaction":
         )
         return cls(raw_transaction_dict["ledger_id"], raw_transaction_dict["body"])
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:97]: no argument types

File: aea/helpers/transaction/base.py [hunk 38]
```diff
@@ -187,7 +185,7 @@ def decode(cls, raw_message_protobuf_object) -> "RawMessage":
             raw_message_dict["is_deprecated_mode"],
         )
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:188]: no argument types

File: aea/helpers/transaction/base.py [hunk 39]
```diff
@@ -362,7 +358,7 @@ def decode(cls, signed_message_protobuf_object) -> "SignedMessage":
             signed_message_dict["is_deprecated_mode"],
         )
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:361]: no argument types

File: aea/helpers/transaction/base.py [hunk 40]
```diff
@@ -433,15 +429,15 @@ def decode(cls, state_protobuf_object) -> "State":
         state_dict = DictProtobufStructSerializer.decode(state_protobuf_object.state)
         return cls(state_dict["ledger_id"], state_dict["body"])
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:432]: no argument types

File: aea/helpers/transaction/base.py [hunk 41]
```diff
@@ -909,7 +905,7 @@ def decode(cls, terms_protobuf_object) -> "Terms":
             **dict(terms_dict["kwargs"]),
         )
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:908]: no argument types

File: aea/helpers/transaction/base.py [hunk 42]
```diff
@@ -1006,15 +1002,15 @@ def decode(cls, transaction_digest_protobuf_object) -> "TransactionDigest":
             transaction_digest_dict["ledger_id"], transaction_digest_dict["body"]
         )
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:1005]: no argument types

File: aea/helpers/transaction/base.py [hunk 43]
```diff
@@ -1095,7 +1093,7 @@ def decode(cls, transaction_receipt_protobuf_object) -> "TransactionReceipt":
             transaction_receipt_dict["transaction"],
         )
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/helpers/transaction/base.py:1096]: no argument types

File: aea/mail/base.py [hunk 44]
```diff
@@ -117,11 +117,11 @@ def port(self) -> Optional[int]:
         """Get the port."""
         return self._port
 
-    def __str__(self):
+    def __str__(self) -> str:
         """Get string representation."""
         return self.uri_raw
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/mail/base.py:124]: no argument types

File: aea/mail/base.py [hunk 45]
```diff
@@ -209,11 +209,11 @@ def _get_public_ids_from_uri(
             )
         return (skill_id, connection_id)
 
-    def __str__(self):
+    def __str__(self) -> str:
         """Get the string representation."""
         return f"EnvelopeContext(connection_id={self.connection_id}, skill_id={self.skill_id}, uri_raw={self.uri_raw})"
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/mail/base.py:216]: no argument types

File: aea/protocols/base.py [hunk 46]
```diff
@@ -240,7 +240,7 @@ def _is_consistent(self) -> bool:  # pylint: disable=no-self-use
         """Check that the data is consistent."""
         return True
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
```
Comment (solarw) [inline aea/protocols/base.py:243]: no argument types

**Review by solarw — APPROVED** [review]
LGTM, some arguments types missing

File: aea/cli/publish.py (hunk 3 above)
Comment (DavidMinarsch) [reply aea/cli/publish.py:132]: done

File: aea/cli/search.py (hunk 6 above)
Comment (DavidMinarsch) [reply aea/cli/search.py:147]: done

File: aea/helpers/multiple_executor.py [hunk 47]
```diff
@@ -205,7 +205,7 @@ async def _wait_tasks_complete(self, skip_exceptions: bool = False) -> None:
         """
         pending = cast(Set[asyncio.futures.Future], set(self._future_task.keys()))
 
-        async def wait_future(future):
+        async def wait_future(future) -> None:  # type: ignore
```
Comment (DavidMinarsch) [inline aea/helpers/multiple_executor.py:208]: There's a problem with this coroutine. @solarw let's discuss.

**Pull request merged**